### DISCUSSION TITLE
How to make a custom control that is a React component & is clickable (v7.x)

### DISCUSSION THREAD

**Discussion by dreamorosi** [uber/react-map-gl#1728] [Q&A] [answered]
Based on my understanding of the documentation the `useControl` hook must be used in conjunction with a JS class that has the [this signature](http://visgl.github.io/react-map-gl/docs/api-reference/use-control#signature) and implements `onCreate` & `onRemove` methods.

The example provided in the repo uses an already-made custom control (`@mapbox/mapbox-gl-draw`) that seems to create its structure by manipulating the DOM directly. If I follow the same type of implementation I'm able to create a custom control just fine.

If instead I want to create a custom control that is also a React component there's no documented / straightforward way that I could find. While looking at the repo I found [this other discussion](https://github.com/visgl/react-map-gl/discussions/1717) that points to the [custom-overlay example](https://github.com/visgl/react-map-gl/tree/7.0-release/examples/custom-overlay).

Following the example just mentioned I am able to render my "control" React component as an overlay using `useControl` & `React.createPortal` on the DOM element created by the custom control class.

The problem with this method is that I'm actually rendering an overlay on the map rather than an actual control and this also means I'm not able to interact with the control at all as there's no way (that I could find) to intercept click events. All events go straight to the map.

My question is:
- Is it possible / in scope to create custom controls that are also React components and are clickable?
- Is the method that I described the correct one (aka rendering the component as a control/overlay)?

As alternative I'm rendering the component outside of the `Map` component (i.e. not as a child of the map) and putting it on the map via CSS but this poses its own challenges with positioning. Aside the fact that it'd be handy to be able to render the element inside the map for positioning reasons, this kind of use case would also allow to create React-based controls instead of plain JS ones.

**Comment by Pessimistress** [comment, marked as answer]
> there's no way (that I could find) to intercept click events

Set the `pointerEvents` CSS.

**Reply by dreamorosi** [reply]
That worked, thanks a lot!

**Reply by ghost** [reply]
@dreamorosi I dont quite understand. What was your final solution? Did you create a JS class with the required methods or a React component?

**Reply by dreamorosi** [reply]
@waltermonecke you need to set `pointerEvents` CSS property on the component that you want to be clickable on the map:
```ts
<Map>
  <MyOverlay
    style={{
      pointerEvents: "all",
    }}
  />
</Map>
```

**Comment by jo-chemla** [comment]
Hey there, 
Thanks for the detailed take on how to add a custom control + answer to catch mouseClicks. I've crawled through the doc and examples, and wanted to make sure this is the simplest way to add a Custom Control Button (or react component) that when clicked, update some state properties. 
Could this be added as a core example in the docs? Or do you have a minimal skeleton for that code? Thanks!

Note: to have the control appear at the correct place, I used the `custom-overlay.tsx` component, and added as a second argument to the `useControl` method the additional parameter `{position: props['position']}`. However, I cannot see to which component I should attribute the className `class="mapboxgl-ctrl mapboxgl-ctrl-group"` so the controls have a consistent mapbox controller style

**Reply by Sonono1103** [reply]
Setting `position` to `getDefaultPosition()` of `OverlayControl` class works fine! Thank you.

I'm not sure if this is the best solution, but `this._container.className = 'mapboxgl-ctrl mapboxgl-ctrl-group'` worked well for me. Also, by doing this, it's no longer necessary to write CSS for pointerEvents. 
Moreover, to apply hover styles, it's necessary to place the icon you want to add under a button like so: `<Overlay><button><svg></svg></button></Overlay>`.

For anyone who might need this, here is my final version of the Overlay code...
```
type Config = {
  position: ControlPosition;
  redraw?: () => void;
};

type OverlayProps = {
  position: ControlPosition;
  children: ReactElement;
};

class OverlayControl implements IControl {
  _map: MapboxMap | null = null;
  _container: HTMLElement | null = null;
  _position: ControlPosition = 'bottom-right';
  _redraw?: () => void;

  constructor({ position, redraw }: Config) {
    this._position = position;
    if (redraw !== undefined) this._redraw = redraw;
  }

  onAdd(map: MapboxMap) {
    this._map = map;
    this._container = document.createElement('div');
    this._container.className = 'mapboxgl-ctrl mapboxgl-ctrl-group';

    if (this._redraw !== undefined) {
      map.on('move', this._redraw);
      this._redraw();
    }

    return this._container;
  }

  onRemove() {
    if (this._map === null || this._container === null) return;
    this._container.remove();
    if (this._redraw !== undefined) this._map.off('move', this._redraw);
    this._map = null;
  }

  getDefaultPosition?() {
    return this._position;
  }

  getMap() {
    return this._map;
  }

  getElement() {
    return this._container;
  }
}

const Overlay = ({ position, children }: OverlayProps) => {
  const ctrl = useControl<OverlayControl>(() => {
    return new OverlayControl({ position });
  });

  const map = ctrl.getMap();
  const elem = ctrl.getElement();
  if (map === null || elem === null) return;

  return createPortal(cloneElement(children, { map }), elem);
};

export default memo(Overlay);

```

Note: An absolutely positioned div doesn't work well for me. When the map size changes, the original icons and the custom icon overlap.

**Reply by david-shortman** [reply]
Great example!

**Comment by kauly** [comment]
In my opinion, it's easier create an absolute positioned div and put whatever component you want inside of it.

**Reply by mattixpet** [reply]
It kinda sucks that that's the case though. But I agree.

**Comment by stychu** [comment]
Sadge there is no easy example how to create custom control.

I've put together something like this and it seems to work pretty well. It respects the ordering of controls and you can put custom react elements as a single element or a group.

```ts
import { useEffect, useState } from 'react';

import { createPortal } from 'react-dom';

import type { PropsWithChildren, ReactNode } from 'react';

type CustomControlProps = {
  position: keyof typeof controlPositions;
};

const controlPositions = {
  topLeft: '.maplibregl-ctrl-top-left',
  topRight: '.maplibregl-ctrl-top-right',
  bottomLeft: '.maplibregl-ctrl-bottom-left',
  bottomRight: '.maplibregl-ctrl-bottom-right',
};

export const CustomControl = ({
  position,
  children,
}: PropsWithChildren<CustomControlProps>) => {
  const [groupContainer, setGroupContainer] = useState<HTMLDivElement | null>(
    null,
  );

  useEffect(() => {
    const parentElement = document.querySelector(controlPositions[position]);
    const groupDiv = document.createElement('div');

    if (parentElement) {
      groupDiv.classList.add('maplibregl-ctrl', 'maplibregl-ctrl-group');

      setGroupContainer(groupDiv);
      parentElement.appendChild(groupDiv);
    }

    return () => {
      if (parentElement) {
        parentElement.removeChild(groupDiv);
      }
    };
  }, []);

  if (!groupContainer) return null;

  return createPortal(<>{children}</>, groupContainer);
};

type DefaultButtonProps = {
  title: string;
  onClick: () => void;
  icon: () => ReactNode;
};

const DefaultButton = ({ title, onClick, icon }: DefaultButtonProps) => {
  return (
    <button type="button" aria-label={title} title={title} onClick={onClick}>
      <span className="flex justify-center items-center" aria-hidden={true}>
        {icon()}
      </span>
    </button>
  );
};

CustomControl.DeafultButton = DefaultButton;

```

And then my custom control for resetting boundaries for example.

```ts
import { useCallback, useEffect } from 'react';

import bbox from '@turf/bbox';
import { RotateCcw } from 'lucide-react';
import { useMap } from 'react-map-gl/maplibre';

import { CustomControl } from '$/components/map/components/CustomControl';

type FitBoundsControlProps = {
  data: Parameters<typeof bbox>[0];
  autoFit?: boolean;
  clickFit?: boolean;
};

export const ResetBoundsControl = ({
  data,
  autoFit = true,
  clickFit = true,
}: FitBoundsControlProps) => {
  const { current: map } = useMap();

  const resetBounds = useCallback(
    (enabled?: boolean) => {
      if (!map || !enabled) return;

      const bounds = bbox(data);

      if (bounds.every((value) => Number.isFinite(value))) {
        const [minLng, minLat, maxLng, maxLat] = bounds;

        map.fitBounds(
          [
            [minLng, minLat],
            [maxLng, maxLat],
          ],
          { padding: 40, duration: 1000 },
        );
      }
    },
    [map, data],
  );

  useEffect(() => {
    resetBounds(autoFit);
  }, [resetBounds]);

  return (
    clickFit && (
      <CustomControl position="topLeft">
        <CustomControl.DeafultButton
          title="Reset Map View"
          icon={() => <RotateCcw />}
          onClick={() => {
            resetBounds(clickFit);
          }}
        />
      </CustomControl>
    )
  );
};

```

```ts
<MapWrapper>
	<FullscreenControl position="top-left" />
      <NavigationControl position="top-left" />
      <ScaleControl />
      <ResetBoundsControl data={geoData} />
</MapWrapper>
```
